src/ucs/sys/socket.c

```diff
@@ -0,0 +1,214 @@
+    }
+
+    status = ucs_netif_ioctl(if_name, SIOCGIFFLAGS, &ifr);
+    if (status != UCS_OK) {
+        return 0;
+    }
+
+    return (ifr.ifr_flags & IFF_UP) && (ifr.ifr_flags & IFF_RUNNING) &&
+           !(ifr.ifr_flags & IFF_LOOPBACK);
+}
+
+ucs_status_t ucs_socket_create(int domain, int type, int *fd_p)
+{
+    int fd = socket(domain, type, 0);
+    if (fd < 0) {
+        ucs_error("socket create failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    *fd_p = fd;
+    return UCS_OK;
+}
+
+ucs_status_t ucs_socket_connect(int fd, const struct sockaddr *dest_addr)
+{
+    char str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+    int ret;
+
+    ret = connect(fd, dest_addr, ucs_sockaddr_sizeof(dest_addr, &status));
+    if ((ret < 0) || (status != UCS_OK)) {
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        if (errno == EINPROGRESS) {
+            ucs_debug("connect(fd=%d, dest_addr=%s): %m", fd,
+                      ucs_sockaddr_str(dest_addr, str, UCS_SOCKADDR_STRING_LEN));
+            return UCS_INPROGRESS;
+        }
+
+        ucs_error("connect(fd=%d, dest_addr=%s) failed: %m", fd,
+                  ucs_sockaddr_str(dest_addr, str, UCS_SOCKADDR_STRING_LEN));
+        return UCS_ERR_UNREACHABLE;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_socket_connect_nb_get_status(int fd)
+{
+    socklen_t conn_status_sz;
+    int ret, conn_status;
+
+    conn_status_sz = sizeof(conn_status);
+
+    ret = getsockopt(fd, SOL_SOCKET, SO_ERROR,
+                     &conn_status, &conn_status_sz);
+    if (ret < 0) {
+        ucs_error("getsockopt(fd=%d) failed to get SOL_SOCKET(SO_ERROR): %m", fd);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    if ((conn_status == EINPROGRESS) || (conn_status == EWOULDBLOCK)) {
+        return UCS_INPROGRESS;
+    }
+
+    if (conn_status != 0) {
+        ucs_error("SOL_SOCKET(SO_ERROR) status on fd %d: %s", fd, strerror(conn_status));
+        return UCS_ERR_UNREACHABLE;
+    }
+
+    return UCS_OK;
+}
+
+size_t ucs_sockaddr_sizeof(const struct sockaddr *addr, ucs_status_t *status_p)
+{
+    ucs_status_t status = UCS_OK;
+    size_t size;
+
+    switch (addr->sa_family) {
+    case AF_INET:
+        size = sizeof(struct sockaddr_in);
+        break;
+    case AF_INET6:
+        size = sizeof(struct sockaddr_in6);
+        break;
+    default:
+        ucs_error("unknown address family: %d", addr->sa_family);
+        status = UCS_ERR_INVALID_PARAM;
+        size = 0;
+        break;
+    }
+
+    if (status_p) {
+        *status_p = status;
+    }
+    return size;
+}
+
+int ucs_sockaddr_get_port(const struct sockaddr *addr)
+{
+    if (addr == NULL)
+        return 0;
+
+    switch (addr->sa_family) {
+    case AF_INET:
+        return ntohs(UCS_SIN_PORT(addr));
+    case AF_INET6:
+        return ntohs(UCS_SIN6_PORT(addr));
+    default:
+        ucs_error("unknown address family: %d", addr->sa_family);
+        return UCS_ERR_INVALID_PARAM;
```

Comment:
0 means any addr for addresses that will be passed to `bind()` and then expect random generated port on which server will listen incoming connection requests.

`int` is not good way here, because `UCS_ERR_INVALID_PARAM` is `ucs_status_t`. Maybe return through optional parameter? WDYT?

Reply:
then, probably, some UCS error code is fine. We can consider the return value to be UCS error code if it s < 0 (like in am_bcopy)

Reply:
thank you!

Reply:
how come the return type of this func is 'int', but we return UCS_ERR here?
need to return the port by reference (e.g add `int *port_p` arg and return ucs_status_t)